src/remote_tables/pushdown_visitor.rs

```diff
@@ -0,0 +1,204 @@
+                // left/right expression is also a BinaryExpr of lower operator precedence.
+                if let Expr::BinaryExpr(right_be @ BinaryExpr { .. }) = &*be.right {
+                    let p = right_be.precedence();
+                    if p == 0 || p < be.precedence() {
+                        right_sql = format!("({})", right_sql)
+                    }
+                }
+                if let Expr::BinaryExpr(left_be @ BinaryExpr { .. }) = &*be.left {
+                    let p = left_be.precedence();
+                    if p == 0 || p < be.precedence() {
+                        left_sql = format!("({})", left_sql)
+                    }
+                }
+
+                let op = self.source.op_to_sql(be.op).ok_or_else(|| {
+                    DataFusionError::Execution(format!(
+                        "Couldn't convert operator {:?} to a compatible one for the remote system",
+                        be.op,
+                    ))
+                })?;
+
+                self.sql_exprs
+                    .push(format!("{} {} {}", left_sql, op, right_sql))
+            }
+            _ => {}
+        };
+        Ok(self)
+    }
+}
+
+// Walk the filter expression AST for a particular remote source type and see if the expression is
+// ship-able, at the same time converting elements (e.g. operators) to the native representation if
+// needed.
+pub fn filter_expr_to_sql<T: FilterPushdownVisitor>(
+    filter: &Expr,
+    source_pushdown: T,
+) -> Result<String> {
+    // Construct the initial visitor state
+    let visitor = FilterPushdown {
+        source: source_pushdown,
+        pushdown_supported: true,
+        sql_exprs: vec![],
+    };
+
+    // Perform the walk through the expr AST trying to construct the equivalent SQL for the
+    // particular source type at hand.
+    let FilterPushdown {
+        pushdown_supported,
+        sql_exprs,
+        ..
+    } = filter.accept(visitor)?;
+
+    if !pushdown_supported || sql_exprs.len() != 1 {
+        return Err(DataFusionError::Execution(format!(
+            "Pushdown not supported for expression {}",
+            filter
+        )));
+    }
+
+    Ok(sql_exprs
+        .first()
+        .expect("Exactly 1 SQL expression expected")
+        .clone())
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::remote_tables::pushdown_visitor::{
+        filter_expr_to_sql, MySQLFilterPushdown, PostgresFilterPushdown,
+        SQLiteFilterPushdown,
+    };
+    use datafusion::logical_expr::{and, col, lit, or, Expr};
+    use rstest::rstest;
+
+    #[rstest]
+    #[case::simple_binary_expression(
+        col("a").gt_eq(lit(25)),
+        "a >= 25")
+    ]
+    #[case::complex_binary_expression(
+        or(and(or(col("a").eq(lit(1)), col("b").gt(lit(10))), col("c").lt_eq(lit(15.0))), col("d").not_eq(lit("some_string"))),
+        "(a = 1 OR b > 10) AND c <= 15 OR d != 'some_string'")
+    ]
```

Comment:
Can you also add a couple of tests here for:

  - strings with single quotes in them
  - timestamps
  - column names with spaces getting quoted properly

(they won't be integration tests but we'll be able to have some cases for them being escaped / serialized correctly)

Reply:
Tests added now.